lib/src/DonutChart.js

```diff
@@ -6,218 +6,156 @@
-import Legend from './Legend';
-
-
-/**
- * Calculates the total of all the items' value,
- * using a reduce function over the items.
- * @param {} items - An array of items, each with values, at least.
- */
-function calculateTotal(items) {
-    return items.reduce((sum, currItem) => sum + currItem.value, 0);
-}
-
-/**
- * @extends {Component}
- */
-export default class DonutChart extends Component {
-    /**
-    * Represents a DonutChart.
-    * @constructor
-    * @param {} props - Object of propTypes with defaultProps as fallback.
-    */
-    constructor(props) {
-        super(props);
-
-        const { data } = props;
-
-        this.state = {
-            selected: data[0],
-            toggleSelect: false
-        };
-    }
-
-    componentWillReceiveProps(newProps) {
-        const { data } = newProps;
-
-        // if new data, reset
-        if (data && (JSON.stringify(data)
-                    !== JSON.stringify(this.props.data))) {
-            this.setState({
-                selected: data[0],
-                toggleSelect: false
-            });
-        }
-    }
-
-    /**
-     * Handles the clicking of an ArcPath region.
-     * Toggles the selected state,
-     * effectively freezing all handlers on other, unselected items.
-     * Calls the onClick handler,
-     * passed by the user as a prop,
-     * providing the item clicked,
-     * as well as whether it's becoming selected or unselected.
-     * @param {} item - The item object selected, with a label and a value, at least.
-     */
-    handleClick(item) {
-        if (this.state.selected.label === item.label) {
-            const toggle = this.props.clickToggle ? !this.state.toggleSelect : false;
-            this.setState({
-                toggleSelect: toggle,
-                selected: item
-            });
-            this.props.onClick(item, toggle);
-        }
-    }
-
-    /**
-     * Handles the mouseenter event over an ArcPath region.
-     * Sets the clicked item as selected in the state object.
-     * @param {} item - The item object selected, with a label and a value, at least.
-     */
-    handleMouseEnter(item) {
-        if (!this.state.toggleSelect) {
-            this.setState({
-                selected: item
-            });
-            this.props.onMouseEnter(item);
+import LegendItem from './LegendItem';
+
+export const DonutChartContext = createContext();
+
+const DonutChart = ({
+  className, clickToggle, colors, colorFunction, data, formatValues, height, legend,
+  onClick, onMouseEnter, width, emptyColor, strokeColor, ...restProps
+}) => {
+  const [selected, setSelected] = useState(data[0]);
+  const [toggleSelect, setToggleSelect] = useState(false);
+
+  useEffect(() => {
+    setSelected(data[0]);
+    setToggleSelect(false);
+  }, [data]);
+
+  const total = data.reduce((sum, currItem) => sum + currItem.value, 0);
+  const graphWidth = legend ? (width * (2 / 3)) : width;
+
+  const itemProps = (item, index) => {
```

Comment:
Shared props for `ArcPaths` and `LegendItems` . Handles selection and toggle-based state changes